Sources/OpenAPIRuntime/Multipart/MultipartSerializer.swift

```diff
@@ -0,0 +1,260 @@
+                        case .start: emitStart()
+                        case .end: emitEnd()
+                        }
+                    }
+                    return flushedBytes()
+                case .emitError(let error): throw SerializerError(error: error)
+                }
+            }
+        }
+    }
+}
+
+extension MultipartSerializer {
+
+    /// An error thrown by the serializer.
+    struct SerializerError: Swift.Error, CustomStringConvertible, LocalizedError {
+
+        /// The underlying error emitted by the state machine.
+        var error: StateMachine.ActionError
+
+        var description: String {
+            switch error {
+            case .noHeaderFieldsAtStart: return "No header fields found at the start of the multipart body."
+            }
+        }
+
+        var errorDescription: String? { description }
+    }
+}
+
+extension MultipartSerializer {
+
+    /// Writes the provided header fields into the buffer.
+    /// - Parameter headerFields: The header fields to serialize.
+    private mutating func emitHeaders(_ headerFields: HTTPFields) {
+        outBuffer.append(contentsOf: ASCII.crlf)
+        let sortedHeaders = headerFields.sorted { a, b in a.name.canonicalName < b.name.canonicalName }
+        for headerField in sortedHeaders {
+            outBuffer.append(contentsOf: headerField.name.canonicalName.utf8)
+            outBuffer.append(contentsOf: ASCII.colonSpace)
+            outBuffer.append(contentsOf: headerField.value.utf8)
+            outBuffer.append(contentsOf: ASCII.crlf)
+        }
+        outBuffer.append(contentsOf: ASCII.crlf)
+    }
+
+    /// Writes the part body chunk into the buffer.
+    /// - Parameter bodyChunk: The body chunk to write.
+    private mutating func emitBodyChunk(_ bodyChunk: ArraySlice<UInt8>) { outBuffer.append(contentsOf: bodyChunk) }
+
+    /// Writes an end of part boundary into the buffer.
+    private mutating func emitEndOfPart() {
+        outBuffer.append(contentsOf: ASCII.crlf)
+        outBuffer.append(contentsOf: ASCII.dashes)
+        outBuffer.append(contentsOf: boundary)
+    }
+
+    /// Writes the start boundary into the buffer.
+    private mutating func emitStart() {
+        outBuffer.append(contentsOf: ASCII.dashes)
+        outBuffer.append(contentsOf: boundary)
+    }
+
+    /// Writes the end double dash to the buffer.
+    private mutating func emitEnd() {
+        outBuffer.append(contentsOf: ASCII.dashes)
+        outBuffer.append(contentsOf: ASCII.crlf)
+        outBuffer.append(contentsOf: ASCII.crlf)
+    }
+}
+
+extension MultipartSerializer {
+
+    /// A state machine representing the multipart frame serializer.
+    struct StateMachine {
+
+        /// The possible states of the state machine.
+        enum State: Hashable {
+
+            /// Has not yet written any bytes.
+            case initial
+
+            /// Emitted start, but no frames yet.
+            case startedNothingEmittedYet
```

Comment:
Needs a better name since the comment suggests emitted _something_ (cf. nothing); specifically the start.

For symmetry with the others, why not, simply `emittedStart`?

Reply:
Sure, good call.

Reply:
https://github.com/apple/swift-openapi-runtime/pull/73/commits/786d00e0dbf1ab040fd59e7e9286bb85b5d193cc